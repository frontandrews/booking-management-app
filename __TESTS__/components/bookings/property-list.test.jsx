import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import PropertyList from '@/components/bookings/property-list';

jest.mock('@/components/bookings/property-dialog', () => ({
  PropertyDialog: () => <div data-testid="property-dialog">PropertyDialog</div>,
}));

jest.mock('@/redux/features/property/slice', () => ({
  setSelectedPropertyId: jest.fn(),
}));

jest.mock('@/redux/features/booking/slice', () => ({
  fetchBookings: jest.fn(),
}));

const createMockStore = (initialState) => {
  return configureStore({
    reducer: {
      property: () => initialState.property,
      booking: () => initialState.booking,
    },
    preloadedState: initialState,
    middleware: (getDefaultMiddleware) => getDefaultMiddleware(),
  });
};

describe('PropertyList Component', () => {
  let store;

  const initialPropertyState = {
    properties: [
      { id: 1, name: 'Property One', location: 'Location One' },
      { id: 2, name: 'Property Two', location: 'Location Two' },
    ],
    selectedPropertyId: 1,
    loading: false,
    error: null,
  };

  const initialState = {
    property: initialPropertyState,
    booking: {},
  };

  beforeEach(() => {
    store = createMockStore(initialState);
  });

  const renderComponent = () =>
    render(
      <Provider store={store}>
        <PropertyList />
      </Provider>,
    );

  test('renders all properties and displays correct count', () => {
    const { container } = renderComponent();

    expect(screen.getByTestId('property-header').textContent).toMatch(
      /Properties\s+\(2\/2\)/,
    );
    expect(screen.getByText('Property One')).toBeInTheDocument();
    expect(screen.getByText('Location One')).toBeInTheDocument();
    expect(screen.getByText('Property Two')).toBeInTheDocument();
    expect(screen.getByText('Location Two')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test('filters properties based on search term', () => {
    const { container } = renderComponent();

    const searchInput = screen.getByPlaceholderText('Search');
    fireEvent.change(searchInput, { target: { value: 'One' } });

    const propertyCount = screen.getByTestId('property-count');
    expect(propertyCount).toHaveTextContent('(1/2)');
    expect(screen.getByText('Property One')).toBeInTheDocument();
    expect(screen.queryByText('Property Two')).not.toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test('displays no results message when search term does not match any property', () => {
    const { container } = renderComponent();

    const searchInput = screen.getByPlaceholderText('Search');
    fireEvent.change(searchInput, { target: { value: 'Three' } });

    expect(
      screen.getByText('No results found for the search criteria.'),
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  test('renders PropertyDialog component for each property', () => {
    const { container } = renderComponent();

    const propertyDialogs = screen.getAllByTestId('property-dialog');
    expect(propertyDialogs.length).toBe(3);

    expect(container).toMatchSnapshot();
  });

  test.todo('handles property click and dispatches actions');
});
