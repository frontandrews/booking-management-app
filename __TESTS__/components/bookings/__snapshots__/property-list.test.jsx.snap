// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyList Component displays no results message when search term does not match any property 1`] = `
<div>
  <div
    class="border p-6"
  >
    <div
      class="flex justify-between items-center flex-row"
    >
      <h1
        class="flex items-center gap-2 text-2xl font-semibold leading-6 py-3 lg:py-6"
        data-testid="property-header"
      >
        Properties
         
        <span
          class="text-sm leading-[0.9] text-gray-500"
          data-testid="property-count"
        >
          (
          0
          /
          2
          )
        </span>
      </h1>
      <div
        class="pb-3 pt-3 pr-0 lg:pr-4"
      >
        <div
          data-testid="property-dialog"
        >
          PropertyDialog
        </div>
      </div>
    </div>
    <div
      class="pb-6 pr-0 lg:pr-4"
    >
      <input
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        placeholder="Search"
        value="Three"
      />
    </div>
    <div>
      <ul
        class="h-[200px] lg:h-[390px] overflow-y-auto pr-1.5 lg:pr-2 custom-scroll"
      >
        <p>
          No results found for the search criteria.
        </p>
      </ul>
    </div>
  </div>
</div>
`;

exports[`PropertyList Component filters properties based on search term 1`] = `
<div>
  <div
    class="border p-6"
  >
    <div
      class="flex justify-between items-center flex-row"
    >
      <h1
        class="flex items-center gap-2 text-2xl font-semibold leading-6 py-3 lg:py-6"
        data-testid="property-header"
      >
        Properties
         
        <span
          class="text-sm leading-[0.9] text-gray-500"
          data-testid="property-count"
        >
          (
          1
          /
          2
          )
        </span>
      </h1>
      <div
        class="pb-3 pt-3 pr-0 lg:pr-4"
      >
        <div
          data-testid="property-dialog"
        >
          PropertyDialog
        </div>
      </div>
    </div>
    <div
      class="pb-6 pr-0 lg:pr-4"
    >
      <input
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        placeholder="Search"
        value="One"
      />
    </div>
    <div>
      <ul
        class="h-[200px] lg:h-[390px] overflow-y-auto pr-1.5 lg:pr-2 custom-scroll"
      >
        <li
          class="flex border justify-between items-center mb-4 p-4 shadow rounded cursor-pointer bg-gray-100 border-"
        >
          <div>
            <h2
              class="text-xl font-bold"
            >
              Property One
            </h2>
            <p>
              Location One
            </p>
          </div>
          <div
            class="flex space-x-4"
          >
            <div
              data-testid="property-dialog"
            >
              PropertyDialog
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`PropertyList Component renders PropertyDialog component for each property 1`] = `
<div>
  <div
    class="border p-6"
  >
    <div
      class="flex justify-between items-center flex-row"
    >
      <h1
        class="flex items-center gap-2 text-2xl font-semibold leading-6 py-3 lg:py-6"
        data-testid="property-header"
      >
        Properties
         
        <span
          class="text-sm leading-[0.9] text-gray-500"
          data-testid="property-count"
        >
          (
          2
          /
          2
          )
        </span>
      </h1>
      <div
        class="pb-3 pt-3 pr-0 lg:pr-4"
      >
        <div
          data-testid="property-dialog"
        >
          PropertyDialog
        </div>
      </div>
    </div>
    <div
      class="pb-6 pr-0 lg:pr-4"
    >
      <input
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        placeholder="Search"
        value=""
      />
    </div>
    <div>
      <ul
        class="h-[200px] lg:h-[390px] overflow-y-auto pr-1.5 lg:pr-2 custom-scroll"
      >
        <li
          class="flex border justify-between items-center mb-4 p-4 shadow rounded cursor-pointer bg-gray-100 border-"
        >
          <div>
            <h2
              class="text-xl font-bold"
            >
              Property One
            </h2>
            <p>
              Location One
            </p>
          </div>
          <div
            class="flex space-x-4"
          >
            <div
              data-testid="property-dialog"
            >
              PropertyDialog
            </div>
          </div>
        </li>
        <li
          class="flex border justify-between items-center mb-4 p-4 shadow rounded cursor-pointer bg-white hover:bg-gray-50"
        >
          <div>
            <h2
              class="text-xl font-bold"
            >
              Property Two
            </h2>
            <p>
              Location Two
            </p>
          </div>
          <div
            class="flex space-x-4"
          >
            <div
              data-testid="property-dialog"
            >
              PropertyDialog
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`PropertyList Component renders all properties and displays correct count 1`] = `
<div>
  <div
    class="border p-6"
  >
    <div
      class="flex justify-between items-center flex-row"
    >
      <h1
        class="flex items-center gap-2 text-2xl font-semibold leading-6 py-3 lg:py-6"
        data-testid="property-header"
      >
        Properties
         
        <span
          class="text-sm leading-[0.9] text-gray-500"
          data-testid="property-count"
        >
          (
          2
          /
          2
          )
        </span>
      </h1>
      <div
        class="pb-3 pt-3 pr-0 lg:pr-4"
      >
        <div
          data-testid="property-dialog"
        >
          PropertyDialog
        </div>
      </div>
    </div>
    <div
      class="pb-6 pr-0 lg:pr-4"
    >
      <input
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        placeholder="Search"
        value=""
      />
    </div>
    <div>
      <ul
        class="h-[200px] lg:h-[390px] overflow-y-auto pr-1.5 lg:pr-2 custom-scroll"
      >
        <li
          class="flex border justify-between items-center mb-4 p-4 shadow rounded cursor-pointer bg-gray-100 border-"
        >
          <div>
            <h2
              class="text-xl font-bold"
            >
              Property One
            </h2>
            <p>
              Location One
            </p>
          </div>
          <div
            class="flex space-x-4"
          >
            <div
              data-testid="property-dialog"
            >
              PropertyDialog
            </div>
          </div>
        </li>
        <li
          class="flex border justify-between items-center mb-4 p-4 shadow rounded cursor-pointer bg-white hover:bg-gray-50"
        >
          <div>
            <h2
              class="text-xl font-bold"
            >
              Property Two
            </h2>
            <p>
              Location Two
            </p>
          </div>
          <div
            class="flex space-x-4"
          >
            <div
              data-testid="property-dialog"
            >
              PropertyDialog
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
